pkg/kubelet/dockershim/checkpoint_store.go

```diff
@@ -0,0 +1,85 @@
+package dockershim
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+// CheckpointStore provides the interface for checkpoint storage backend
+type CheckpointStore interface {
+	Add(key string, data []byte) error
+	Get(key string) ([]byte, error)
+	Delete(key string) error
+	List() ([]string, error)
+}
+
+// FileStore is an implementation of CheckpointStore interface which stores checkpoint in file.
+type FileStore struct {
+	path string
+}
+
+func (fstore *FileStore) Add(key string, data []byte) error {
```

Comment:
Checking for valid path is an OK start, but insufficient.  I can still pass "." or ".." and totally hork you.  Let me make an argument for being much tighter than that.  As it is, almost any string is a valid name.  This means that there are probably unknown-unknowns lurking (do you handle NUL or utf-8 or ...), but more importantly it means you have no reserved space to work - e.g. your tempfiles can always be in conflict.

I'd argue to spec and enforce this as a) not starting with a dot (so you can use .foo.tmp and never have a conflict) and b) matching a regex.  In fact, you can even go so far as to use the existing validation code like `IsDNS1123Subdomain`.  You can always loosen it if you have need, but locking it down later is harder.